### PR TITLE
feat(mcp): the OAuth surface - DCR, PKCE authorize, consent and token exchange

### PR DESCRIPTION
**Pull request by mosamlife**
Stacked on #572 (S6a-q query layer). **Base is `feat/adr064-s6a-q-mcp-query-layer`, not `main`.** Draft.

DCR (RFC 7591), the authorization endpoint with PKCE, the consent screen's server side, the token endpoint, and the four tx helpers.

## The four tx helpers

`internal/db/db.go` gains `InMCPClientRegisterTx`, `InMCPClientLookupTx`, `InMCPCodeLookupTx`, `InMCPTokenLookupTx` — four rather than one `InMCPTx`, because each opens exactly one table and the narrow name at the call site is what tells you which table you may touch. They share an unexported body so no call site can invent a fifth GUC in passing.

`repo.go` pairs every query with its helper and states the trap at both places it could be got wrong: the lookup policies are `FOR SELECT`, so an UPDATE inside one matches zero rows and raises no error under `FORCE ROW LEVEL SECURITY`. Authentication is two queries, not a join — `mcp_grants` has no lookup policy, so `token JOIN grant` inside the token-lookup transaction matches zero rows and would fail every MCP request silently.

## Proofs — each planted, watched to fail, restored

**Revocation.** Replacing the refusal with `_ = chk.Authorized` (branch on row presence, the stated bug):

```
--- FAIL: TestAuthenticate_RevocationBitesOnTheNextRequest
    a REVOKED grant was still honoured on the next request; revocation must not wait for token expiry
--- FAIL: TestAuthenticate_BranchesOnAuthorizedNotOnRowPresence
    a row with Authorized=false was honoured; the service is branching on row presence
FAIL  github.com/mosamlife/wpmgr/apps/api/internal/mcp  0.631s
```
exit 1. Restored → exit 0. The token in that test is deliberately unexpired, so a service that only checked expiry would pass; this catches it.

**Single use — and this one found a hole in my own proof.** Planting "treat `pgx.ErrNoRows` from the consume as already fine" **did not** redden the replay tests. Both were being refused earlier by the lookup's `IsConsumed` column and never reached the compare-and-set at all — they would have passed with the `ErrNoRows` branch deleted entirely. That is exactly the defect class this slice is guarding against, sitting inside the test written to catch it.

The fake now models the TOCTOU window the compare-and-set exists to close: `raceLost` makes the lookup report a healthy, redeemable code while the UPDATE matches nothing, which is the only path that exercises that branch. With the defect planted:

```
--- FAIL: TestExchange_LosingCompareAndSetIsRefusedNotShrugged
    the losing exchange was granted {AccessToken:D_Zqft8FMKuV3DxjYiiCdBHHRgYq8GWaR2n91EYKCcU
    TokenType:Bearer ExpiresIn:7776000 Scope:mcp:read}; a zero-row consume means REFUSE,
    never 'already fine'
```
exit 1 — a real access token minted for the loser of the race. Restored → exit 0, one consume attempted, zero tokens minted. The test also asserts `consumeCalls == 1` so it cannot silently go back to passing on an earlier check.

**The exit gate now runs end to end through the real Gin handler**, not just the parser: six absence/unrecognised cases each get 400 `invalid_scope` in the RFC 6749 envelope, and the test asserts no consent context leaks on refusal. The over-fire case still returns 200 with the scope granted.

## Other obligations carried

- `redirect_uri` is **exact-matched in Go** against the stored array. Eight redirector shapes are refused by test — path traversal, suffix extension (`...auth_callback.evil.com`), query-appended, subdomain (`claude.ai.evil.com`), truncation, scheme downgrade. There is no `redirect_uri` parameter on the client lookup, so this cannot drift into SQL.
- The consent DTO names the identity `client_name_unverified` and carries `identity_verified` as a **literal `false`** with no field able to set it true, plus `redirect_host` — the part a human can actually judge.
- PKCE mandatory, S256 only; an absent method does not decay to `plain`.
- An empty resolved site scope produces a `SiteSet` that allows nothing (obligation 2), proven through the real service path with a tag matching no site.
- A refused exchange never consumes the code, so a bad request cannot burn a legitimate one.

## Gates, each run unpiped

`go build ./...` exit 0 (pre-existing `ld:` alignment warning from `cmd/media-encoder`), `go vet ./...` exit 0, `go test ./internal/...` exit 0.

## Not here

No MCP protocol, no tool registry, no read tools — that is S6b-2. `Service.Authenticate` is the per-request gate those tools sit behind. The routes are **not mounted in `server.go`** yet; mounting belongs with the transport that uses them.

---

## Inherited debt: this package has NO integration coverage

`grep -rl 'mcp' apps/api/tests/*.go` returns nothing. The integration suite cannot reach `Approve`, `Exchange` or `Authenticate`, so **every proof on this PR is against `fakeStore`, and none runs as the `wpmgr_app` role.**

That matters specifically because the security boundary here is RLS-shaped and the fake cannot model it:

- The four GUC helpers (`InMCPClientRegisterTx`, `InMCPClientLookupTx`, `InMCPCodeLookupTx`, `InMCPTokenLookupTx`) are verified **by reading `repo.go`**, not by executing them. Nothing has yet proven that a write inside a `…ForLookup` transaction really does match zero rows as `wpmgr_app` — the exact failure m124 obligation 1 says is silent.
- `ResolveMCPGrantScopeSitesInTenantTx` dropping foreign UUIDs depends on `sites` RLS. The fake returns whatever it is handed, so the empty-set proof tests the Go half only.
- B1's fix (`LookupClient` + `exactMatchRedirectURI` in `Approve`) is proven through the real Gin handler but against the fake store.

**Whoever mounts these routes inherits the obligation** to add a tenant fixture and integration tests covering: the consume actually writing as `wpmgr_app`, a lookup-transaction write matching zero rows, and cross-tenant site resolution. That belongs with S6b-2's transport, since the routes are not mounted in `server.go` yet.

## Summary by CodeRabbit

* **New Features**
  * Added OAuth support for MCP clients, including dynamic registration, authorization, consent, and token exchange.
  * Added PKCE protection, redirect URI validation, client authentication, token issuance, and single-use authorization codes.
  * Added authenticated access checks with tenant- and site-scoped permissions.
  * Added OAuth-compliant responses, errors, and cache-control headers.

* **Bug Fixes**
  * Improved registration reliability by validating stored client data before reporting success.
  * Ensured failed token exchanges do not consume authorization codes.

* **Tests**
  * Added comprehensive coverage for security, validation, authorization, registration, and end-to-end OAuth flows.

### CODE REVIEW

**Comment by mosamlife** [comment]
Review fixes pushed in `e84bdfa0`. B1, H2 and H3 all addressed; every fix has the mutation that proves its test fires.

## B1 — fixed

`Approve` now re-resolves the client by `client_id` and exact-matches the redirect against the **stored** array before anything is minted. Grant and code record the resolved client id, not the submitted one.

Mutation (removed the lookup and the match, rebuilt `ConsentContext` from the body as before) — RED:

```
TestApprove_RefusesARedirectTheClientNeverRegistered/https://attacker.example/exfiltrate
  Approve minted code "UlrZRF9pqHeF6rEBHI6VNjX5esqgqlFfIEy1v6DQOQs" for unregistered
  redirect_uri "https://attacker.example/exfiltrate"
TestApprove_RefusesAClientThatWasNeverRegistered
  Approve minted code "oBhfwy5x72M0iGu66X73GGB9_-goQpWQOuWFDnp9bVI" for an unregistered client_id
TestConsentHandler_RefusesAnUnregisteredRedirect
  POST /consent returned 200 ... {"grant_id":"49a5e6ff-...","code":"sD1Vc5Zz...",
  "redirect_uri":"https://attacker.example/exfiltrate","state":"s"}
```

Seven attacker redirects, each minting a real code. Restored → exit 0, and the tests assert no grant was created, since a minted grant is not undone by refusing later. `TestApprove_ResolvesTheClientBeforeMinting` pins the mechanism so the fix cannot regress into comparing the body against itself.

## H2 — the approval path now has tests

`approve_test.go`. Each of your five green mutations was re-planted and now reddens: **G** (six scope shapes), **H** (five challenge shapes incl. plain downgrade), **J**, **L**, **A**.

**J is worth flagging: my first version of that test passed with the defect planted.** It stored the raw verifier as the challenge, so it was refused by the PKCE mismatch rather than by the method check — the same "passing for a reason other than the one it claims" shape as the replay-test hole in the previous commit. The challenge is now left as the correct S256 value so only the method can refuse it. Second time this exact shape has appeared in my own tests; the plant-and-watch step is the only thing catching it.

## H3 — implemented, not documented away

`Exchange` now requires `client_id` unconditionally (it was `if req.ClientID != "" && ...`, so omitting it skipped the code-to-client binding — absence read as permission), resolves the client, and for any method other than `none` requires the secret and compares it to the stored hash in constant time. HTTP Basic per RFC 6749 §2.3.1 wins over the body. A non-`none` client with a NULL hash is refused rather than read as "no secret required".

The comment now states the mode that exists, and records that it was false when written so removing the check means removing the comment.

## What is NOT proven — please weigh this

**I have no `wpmgr_app` proof of the B1 fix, and I am not claiming one.** `grep -rl 'mcp' apps/api/tests/*.go` returns nothing — the integration suite contains zero references to this package, so `make test-integration` would not exercise `Approve` at all. Reporting it green would be a check that does not cover the thing it is offered as proof of.

Separately: **my previous report of the integration suite was wrong.** I ran it as `make test-integration 2>&1 | tail -30`, which truncated the `--- FAIL:` lines and gave me the pipe's exit status, so I reported "no result" when the run had in fact failed (`make: *** [test-integration] Error 1`, `FAIL github.com/mosamlife/wpmgr/apps/api/tests 993.465s`). I have not yet established whether that failure is pre-existing; this branch mounts no routes, so it is likely unrelated, but I have not proven that and a second suite (PID 80588) is running now, so I have not re-run.

Real coverage for this path needs the routes mounted and a tenant fixture — that belongs with S6b-2's transport, and I would rather flag it than fake it.

**Comment by mosamlife** [comment]
## Integration suite: GREEN on this branch. The earlier failure was contention.

Run with the machine to itself, on `feat/adr064-s6b1-mcp-oauth-surface` @ `e84bdfa0`, clean tree:

```
make -C <worktree> test-integration > /tmp/s6b1-integ-branch.log 2>&1
rc=$?          # redirect, not a pipe -- rc is make's own status
```

```
BRANCH_EXIT=0
15:12:13Z -> 15:28:43Z   (990s)

ok  github.com/mosamlife/wpmgr/apps/api/tests            985.388s
ok  github.com/mosamlife/wpmgr/apps/api/tests/contract     3.359s
ok  github.com/mosamlife/wpmgr/apps/api/tests/gh458       10.076s
ok  github.com/mosamlife/wpmgr/apps/api/tests/testinfra    2.218s
```

`grep -c '^--- FAIL' /tmp/s6b1-integ-branch.log` -> `0`.

No merge-base run was needed: the matched sample only separates "this branch broke it" from "already red" when the branch is red, and it is not.

### What the earlier failure actually was

My previous run (`FAIL .../apps/api/tests 993.465s`) overlapped another worktree's suite. Contention was confirmed twice while waiting for this one: PID 80588 held the machine for ~14:50, and my bounded waiter cleared only after 14 polls. **This is further evidence for GH #565**, which already records this suite failing roughly 2-in-3 under concurrent runs with `port "5432/tcp" not found`. The suite spins a fresh Postgres container per test, so two runs contend for ports.

Practical consequence: a red `make test-integration` on this repo is not evidence of a defect until it has been re-run alone. Checking `ps -eo pid,etime,command | grep '[t]ests.test'` first costs nothing and prevents a false regression report.

### One more instance of the same defect class, in my own harness

The wrapper script reported **"failed with exit code 1"** while `BRANCH_EXIT=0` sat inside its own output. The script's last statement was `grep -c '^--- FAIL' "$LOG"`, and `grep -c` exits 1 when it matches nothing — so *finding zero failures* made the job look like a failure. The script's status came from the wrong command, exactly as `| tail -30` did earlier, just inverted: last time a real failure read as success, this time a real success read as failure.

It was caught only because the true status was captured explicitly into `BRANCH_EXIT=` and read from there rather than trusting the job's exit code. That is the third instance of this shape in this slice, and the second inside tooling written specifically to avoid it.

File: apps/api/internal/mcp/service.go [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+		return IssuedToken{}, fmt.Errorf("consume authorization code: %w", err)
+	}
+
+	// 6. Only now does a credential exist.
+	secret, err := randomToken(32)
+	if err != nil {
+		return IssuedToken{}, fmt.Errorf("generate connection token: %w", err)
+	}
+	expiresAt := s.now().UTC().Add(connectionTokenTTL)
+	if _, err := s.store.CreateConnectionToken(ctx, sqlc.CreateMCPConnectionTokenParams{
```
[the remark this answers is not in the document]
Comment (mosamlife) [reply apps/api/internal/mcp/service.go]: Valid — fixed in `6dd53ba6`. You are right that the ordering is correct and I have not flipped it; the defect was the two commits, not the sequence.

`Repo.ConsumeAuthorizationCode` and `Repo.CreateConnectionToken` are replaced by a single `RedeemAuthorizationCode` that runs the compare-and-set and the token insert in **one `InTenantTx`**. Either both land or neither does. There is deliberately no longer any way to perform half of it, so a future caller cannot reintroduce the window.

Two properties are unchanged and worth stating, because they are the reasons the shape was the way it was:

- The consume still runs in a transaction **separate from the lookup** — that separation is m124 obligation 1 (the lookup policy is `FOR SELECT`, so a write there matches zero rows silently). What merged is the consume with the *insert*, not the consume with the *lookup*.
- The compare-and-set still runs **first inside** the transaction, so two concurrent exchanges still yield exactly one winner; the loser's INSERT never executes because its UPDATE matched nothing.

A transient failure now also surfaces as an infra error rather than a domain refusal, so a client reads "retry this" rather than `invalid_grant`, which would send it back through the whole browser flow.

### Proof, red before green

The fake models the **rollback**, not the happy path: when the insert fails it does not flip `consumed`, exactly as the real transaction rolls the UPDATE back. That mattered — planting the two-commit behaviour (`f.consumed = true` before returning the error) reddens it:

```
--- FAIL: TestExchange_TokenPersistenceFailureLeavesTheCodeRedeemable
    the authorization code was left CONSUMED after the token insert failed. It can
    never be redeemed, the client did nothing wrong, and it must restart the entire
    browser flow. Consume and issue must share one transaction
```

Restored → green. The test also proves the code is genuinely **redeemable afterwards** by retrying the same request against a healthy store and asserting exactly one token is minted — a rollback-less fake cannot satisfy that half, which is what stops this test passing for the wrong reason.

File: apps/api/internal/mcp/handler.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+		return
+	}
+
+	// RFC 6749 2.3.1 prefers HTTP Basic for client_secret_basic and permits the
+	// body for client_secret_post. Basic wins when both are present, so a body
+	// parameter cannot override a header the client actually authenticated with.
+	clientID, clientSecret := body.ClientID, body.ClientSecret
+	if id, secret, ok := c.Request.BasicAuth(); ok {
+		clientID, clientSecret = id, secret
+	}
```
[the remark this answers is not in the document]
Comment (mosamlife) [reply apps/api/internal/mcp/handler.go:194]: Valid — fixed in `6dd53ba6`. **I went with enforcing it**, and the RFC supports that reading rather than merely permitting it.

RFC 6749 §2.3.1 says a server **must** support Basic and **may** support the body form. That governs what a server has to *offer*; it does not say a server must accept whichever mechanism turns up for a client that already declared one. The same section also says "the client MUST NOT use more than one authentication method in each request", which only means something if the server is paying attention to which method was used — so the spec already assumes the server tracks the transport.

The decisive argument is the column. `token_endpoint_auth_method` is `NOT NULL` over a closed set with no default, so the value is *always* a decision somebody made at registration. Storing a decision and not honouring it is the same shape as a column no statement can write — which m126 deleted rather than leave lying around looking like it governed something. And the strict reading costs a legitimate client nothing, because the client chose the value itself.

What changed:

- The handler previously let Basic **silently win** when both were present, discarding the credential source entirely. It now derives the transport and passes it to the service, which requires it to equal the registered method.
- Presenting credentials through **both** mechanisms is refused rather than resolved by precedence, per the MUST NOT above. That check moved up with the request-shape validations, since it is a property of the request and not of any particular code — a malformed request should not need a valid code to be rejected.
- A client registered `none` presenting a secret is refused too. There is no stored hash to compare against, so accepting one would be accepting a credential nothing verifies — the absence-means-permitted reflex m124 DECISION 11 exists to prevent.

### Proof, red before green

Disabling the transport equality check reddens three cases that must never pass:

```
--- FAIL: .../basic_client_posting_body_credentials
    a client registered client_secret_basic authenticated via client_secret_post;
    the stored method governs nothing
--- FAIL: .../post_client_using_HTTP_Basic
    a client registered client_secret_post authenticated via client_secret_basic;
    the stored method governs nothing
--- FAIL: .../basic_client_presenting_nothing
```

Restored → green, and the two matching cases (basic-with-basic, post-with-post) pass throughout, so the rule does not redden a client using its own registered method. The double-presentation case is driven end to end through the real Gin handler, since only the handler can observe the transport at all.

File: apps/api/internal/mcp/oauth_test.go [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+			q.Set("client_id", "client-abc")
+			q.Set("redirect_uri", redirect)
+			q.Set("code_challenge", "abc123challenge")
+			q.Set("code_challenge_method", "S256")
+			if tc.scope != "" {
+				q.Set("scope", tc.scope)
+			}
+
+			w := httptest.NewRecorder()
+			req := httptest.NewRequest(http.MethodGet, "/api/v1/oauth/mcp/authorize?"+q.Encode(), nil)
```
[the remark this answers is not in the document]
Comment (mosamlife) [reply apps/api/internal/mcp/oauth_test.go:278]: Not adopting this, and the reason is repo-wide consistency rather than disagreement with the idea.

**The linter is not configured and is not run anywhere in this repository.** Verified rather than assumed:

```
find . -name '.golangci*'                     -> no matches
grep -rln 'golangci\|noctx' .github/workflows -> no matches
```

So the `golangci-lint (2.12.2)` diagnostic in the finding comes from CodeRabbit's own sandbox with default linters, not from anything this project runs. `ci.yml` gates on `go build`, `go vet` and `go test`.

**The convention datum is the decisive one:**

```
grep -rn 'httptest.NewRequest('            --include='*.go' . | wc -l  -> 193
grep -rn 'httptest.NewRequestWithContext(' --include='*.go' . | wc -l  ->   0
```

Converting the 4 sites in this package would make `internal/mcp` the only one of 193 using a different form, leaving 189 unconverted. That is a repo-wide convention change, and an OAuth PR is the wrong place to make it in passing.

**On whether there is a real point underneath the linter advice** — I checked rather than dismissing it, because "no linter runs this" is not a reason to leave genuinely worse code. There is no behavioural difference here. All four call sites drive a Gin handler backed by `fakeStore`: no outbound I/O, no goroutines, no timers, and nothing anywhere in the path observes cancellation. `t.Context()` would improve matters where a request context can actually be cancelled and something acts on it; that is not this. The handler does read `c.Request.Context()`, but only to pull the principal the test middleware injected.

Worth noting the codebase already uses `t.Context()` heavily (1017 occurrences) — it is specifically the `httptest` constructor pairing that is uniform at 193/0, which is what makes this a convention question rather than an oversight.

If the project does want `noctx`, the right shape is a `.golangci.yml` plus a CI job plus one sweep across all 193 sites, as its own change with its own review. Happy to see that raised as an issue; it should not ride along here.

Resolving on that basis.
